rlmolecule/mcts/mcts.py

```diff
@@ -0,0 +1,183 @@
+        action_selection_function = self.softmax_selection if explore else self.visit_selection
+        num_mcts_samples = self._num_mcts_samples if num_mcts_samples is None else num_mcts_samples
+
+        def ucb_selection(parent):
+            return max(parent.children, key=lambda child: self._ucb_score(parent, child))
+
+        mcts_selection_function = ucb_selection if mcts_selection_function is None else mcts_selection_function
+
+        return self.run_from_vertex(vertex, action_selection_function, num_mcts_samples, mcts_selection_function)
+
+    def run_from_vertex(
+            self,
+            vertex: MCTSVertex,
+            action_selection_function: Callable[[MCTSVertex], MCTSVertex],
+            num_mcts_samples: int,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> []:
+        path: [] = []
+        while True:
+            # path.append(vertex)
+            self.sample(vertex, num_mcts_samples, mcts_selection_function)
+            self._accumulate_path_data(vertex, path)
+            children = vertex.children
+            if children is None or len(children) == 0:
+                break
+            vertex = action_selection_function(vertex)
+
+        return path
+
+    def sample(
+            self,
+            vertex: MCTSVertex,
+            num_mcts_samples: int,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> None:
+        for _ in range(num_mcts_samples):
+            search_path, value = self._select(vertex, mcts_selection_function)
+            self._backpropagate(search_path, value)
+
+    # noinspection PyMethodMayBeStatic
+    def _accumulate_path_data(self, vertex: MCTSVertex, path: []):
+        path.append(vertex)
+
+    def _select(
+            self,
+            root: MCTSVertex,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> ([MCTSVertex], float):
+        """
+        Selection step of MCTS
+        From Wikipedia (https://en.wikipedia.org/wiki/Monte_Carlo_tree_search):
+        Selection: Start from root R and select successive child vertices until a leaf vertex L is reached.
+        The root is the current game state and a leaf is any vertex that has a potential child from which no simulation
+        (playout) has yet been initiated. The section below says more about a way of biasing choice of child vertices that
+        lets the game tree expand towards the most promising moves, which is the essence of Monte Carlo tree search.
+        """
+        current = root
+        search_path = []
+        while True:
+            search_path.append(current)
+            children = current.children
+
+            if children is None:  # node is unexpanded: expand and return its value estimate
+                return search_path, self._evaluate(current, search_path)
+
+            if len(children) == 0:  # node is expanded and terminal: return its value
```

Comment:
We use `len(children) == 0` a lot, is there any reason we don't want a `State.terminal` property?

Reply:
MCTSVertex::terminal would be what you would want. I have no objection to adding the property, but I didn't see the need. In these cases we are also using children for other purposes. In these cases, it is likely more performant and I believe less error prone to only access the children once, and to use len(children) == 0 explicitly, which is why I wrote it this way. However, it's a point of style more than anything. I am happy to use a property instead. Let me know what you would like.

Reply:
I can see that now, I didn't appreciate that state.terminal was calling `get_next_actions` a bunch of extra times